### PR TITLE
Fix import mechanism for task modules.

### LINKED ISSUE #374: BUG: module.__name__ not set set correctly during exec in test collection
- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of pytask.
- [x] (optional) I have confirmed this bug exists on the `main` branch of pytask.

---

#### Problem description

First of all, thanks for this library! A lot of great stuff in here.

See #373 for a PR that adds a failing test.

While I'm not sure what the root cause is / who's fault it is, the actual method of failure is something like this:
Say I have a file `task_a.py`. In https://github.com/pytask-dev/pytask/blob/9b2d93ee103761e85ef77aa606d1d775fcb19b6b/src/_pytask/collect.py#L108-L120, the module spec is created with a module name of `path.stem`, so in my case it is `task_a`. But then IFF you have used a `from __future__ import annotations`, then you go down this code path where dataclasses tries to call ` sys.modules.get("task_a")`, and "task_a" is not importable by itself.

I think we need to be setting the module.__name__ to something else, so it actually is findable within sys.modules, or import/exec it slightly differently?

### PR DESCRIPTION
**Pull request by NickCrews**
Closes #374.

Corresponding issue in pytest: `https://github.com/pytest-dev/pytest/issues/7856`

I get this obscure error from within dataclasses:

```
────────────────────────── Failures during collection ──────────────────────────
───────── Could not collect test_collect_module_name_0/task_module.py ──────────

╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ in exec_module:883                                                           │
│ in _call_with_frames_removed:241                                             │
│                                                                              │
│ /private/var/folders/nz/wfwcn6md1796t48tx9dhf0v40000gn/T/pytest-of-nickcrews │
│ /pytest-41/test_collect_module_name_0/task_module.py:7 in <module>           │
│                                                                              │
│   4 import dataclasses                                                       │
│   5                                                                          │
│   6 @dataclasses.dataclass                                                   │
│ ❱ 7 class Data:                                                              │
│   8 │   x: int                                                               │
│   9                                                                          │
│                                                                              │
│ /Users/nickcrews/.pyenv/versions/3.10.4/lib/python3.10/dataclasses.py:1263   │
│ in dataclass                                                                 │
│                                                                              │
│   1260 │   │   return wrap                                                   │
│   1261 │                                                                     │
│   1262 │   # We're called as @dataclass without parens.                      │
│ ❱ 1263 │   return wrap(cls)                                                  │
│   1264                                                                       │
│   1265                                                                       │
│   1266 def fields(class_or_instance):                                        │
│                                                                              │
│ /Users/nickcrews/.pyenv/versions/3.10.4/lib/python3.10/dataclasses.py:1253   │
│ in wrap                                                                      │
│                                                                              │
│   1250 │   """                                                               │
│   1251 │                                                                     │
│   1252 │   def wrap(cls):                                                    │
│ ❱ 1253 │   │   return _process_class(                                        │
│   1254 │   │   │   cls, init, repr, eq, order, unsafe_hash, frozen, match_ar │
│   1255 │   │   )                                                             │
│   1256                                                                       │
│                                                                              │
│ /Users/nickcrews/.pyenv/versions/3.10.4/lib/python3.10/dataclasses.py:1003   │
│ in _process_class                                                            │
│                                                                              │
│   1000 │   │   # See if this is a marker to change the value of kw_only.     │
│   1001 │   │   if _is_kw_only(type, dataclasses) or (                        │
│   1002 │   │   │   isinstance(type, str)                                     │
│ ❱ 1003 │   │   │   and _is_type(type, cls, dataclasses, dataclasses.KW_ONLY, │
│   1004 │   │   ):                                                            │
│   1005 │   │   │   # Switch the default to kw_only=True, and ignore this     │
│   1006 │   │   │   # annotation: it's not a real field.                      │
│                                                                              │
│ /Users/nickcrews/.pyenv/versions/3.10.4/lib/python3.10/dataclasses.py:764 in │
│ _is_type                                                                     │
│                                                                              │
│    761 │   │   if not module_name:                                           │
│    762 │   │   │   # No module name, assume the class's module did           │
│    763 │   │   │   # "from dataclasses import InitVar".                      │
│ ❱  764 │   │   │   ns = sys.modules.get(cls.__module__).__dict__             │
│    765 │   │   else:                                                         │
│    766 │   │   │   # Look up module_name in the class's module.              │
│    767 │   │   │   module = sys.modules.get(cls.__module__)                  │
╰──────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'NoneType' object has no attribute '__dict__'

```

#### Changes

Provide a description and/or bullet points to describe the changes in this PR.

#### Todo

- [ ] Reference issues which can be closed due to this PR with "Closes #x".
- [ ] Review whether the documentation needs to be updated.
- [ ] Document PR in docs/changes.rst.

### CODE REVIEW

**Comment by tobiasraabe** [comment]
Hi @NickCrews, do the changes solve your issue?

**Comment by NickCrews** [comment]
Thanks for the quick turnaround! This fixes that exact problem, but introduces a new one:

Say I have `mypkg/io.py`. Then this will set `sys.modules["io"] = <my module>` shadowing the builtin module `io`, and then any other code that does `import io` will import the wrong thing and break.

I'm not exactly sure, but should we be doing [what pytest does](https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html#test-modules-conftest-py-files-inside-packages), and search up the directory tree until we don't find an `__init__.py`, and add that root dir to `sys.path`, then `mypkg/io.py` would be imported as `mypkg.io` (which is what the rest of my application code does, which is correct).

**Comment by tobiasraabe** [comment]
You are right. This was too quick. Thanks for the comment.

To comment on your example with the `io` module, it would only concern `task_modules`, and I would consider it best practices to not import from task modules because it can easily lead to circular import errors. (I should check again whether it is documented.)

The approach implemented in pytask is closest to the `importlib` import mode. This is also the import mode pytest [recommends for new projects](https://docs.pytest.org/en/7.3.x/explanation/goodpractices.html#tests-outside-application-code) and [drawbacks and advantages are also discussed here](https://docs.pytest.org/en/7.3.x/explanation/goodpractices.html#choosing-an-import-mode).

The major drawback that I see is that task modules need to have unique names like `task_module` or they need to have unique names based on the package they are embedded in, like `project.data_management.task_module`. I assume this is not a real problem for existing pytask projects since the recommendation is to create projects based on the src layout.

The benefit of this approach is that we do not need to change `sys.path` at all, which introduces a lot of [complicated logic](https://github.com/pytest-dev/pytest/blob/07eeeb8dfc82de289565a57880ca08b0eac4462d/src/_pytest/pathlib.py#L541-L589).

If we follow this approach, I think three things must be done.
- Add the `root_path` argument to `import_path`: [link](https://github.com/pytest-dev/pytest/blob/07eeeb8dfc82de289565a57880ca08b0eac4462d/src/_pytest/pathlib.py#L489)
- Create the module name from the path: [link](https://github.com/pytest-dev/pytest/blob/07eeeb8dfc82de289565a57880ca08b0eac4462d/src/_pytest/pathlib.py#L524)
- Insert missing modules: [link](https://github.com/pytest-dev/pytest/blob/07eeeb8dfc82de289565a57880ca08b0eac4462d/src/_pytest/pathlib.py#L538)

@NickCrews, how does it sound to you? Since you mentioned the `prepend` and `append` modes, do you see a use case where the `importlib` mode does not achieve what you want and where we need more flexibility?

**Comment by NickCrews** [comment]
Thanks for the patience, and thanks for going ahead and implementing something concrete. I was having trouble understanding exactly what you were proposing until I read those commits.

 I'll leave a few nits as review but in general looks good, thank you!

> The major drawback that I see is that task modules need to have unique names like task_module

You're saying this is the drawback of `prepend` and `append`, right? Because this is NOT a requirement for `importlib`, what you have implemented here. I think you even have a test for this?

> best practices to not import from task modules because it can easily lead to circular import errors

I respectfully disagree. I would like to have my tasks mixed with my "main logic" code. I have a layout where each .py files represents one discrete step in processing. I want each step to export 3 different interfaces:
1. A public python API, eg `def clean_names(names: pd.DataFrame) -> pd.DataFrame`
2. A task that wraps this python API eg `def task_clean_names(....)`, so that this step can be orchestrated with all the other steps in a larger pipeline.
3. A CLI to run this single step of processing, which loads/saves based on the paths specified in `tast_clean_names`

I want all three of these related functionalities to be in one file. This requires that one module containing tasks needs to import other modules containing tasks. Now, this actually works fine right now and I have no complaints! I just need to import as `from mypkg.names.clean import clean_names` and then make sure that `mypkg` is installed/discoverable, but this is standard stuff outside of pytask. But I just wanted to flag that I don't think we should assume/advocate for keeping tasks outside of app code. IDK, maybe I just haven't run into this circ dependency issue yet.

I think avoiding adjusting `sys.path` is a very good goal.

This implementation worked fine for my project. I could see this still not covering someone else's use case, but I think this is at least moving in the right direction, and I don't think should break anyone (unless they are relying on something weird/fragile with the value of `__name__`).

File: src/_pytask/path.py [hunk 1]
```diff
@@ -120,3 +123,76 @@ def find_case_sensitive_path(path: Path, platform: str) -> Path:
...
+    sys.modules[module_name] = mod
+    spec.loader.exec_module(mod)
+    _insert_missing_modules(sys.modules, module_name)
+    return mod
+
+
+def _module_name_from_path(path: Path, root: Path) -> str:
+    """Return a dotted module name based on the given path, anchored on root.
+
+    For example: path="projects/src/tests/test_foo.py" and root="/projects", the
```
Comment (NickCrews) [inline src/_pytask/path.py:153]: Adjust test -> task

File: src/_pytask/path.py [hunk 2]
```diff
@@ -120,3 +123,76 @@ def find_case_sensitive_path(path: Path, platform: str) -> Path:
...
+        # Use the parts for the relative path to the root path.
+        path_parts = relative_path.parts
+
+    return ".".join(path_parts)
+
+
+def _insert_missing_modules(modules: dict[str, ModuleType], module_name: str) -> None:
+    """Insert missing modules when importing modules with :func:`import_path`.
+
+    When we want to import a module as ``src.tests.test_foo`` for example, we need to
```
Comment (NickCrews) [inline src/_pytask/path.py:174]: test -> task

File: tests/test_path.py [hunk 3]
```diff
@@ -117,3 +123,179 @@ def test_find_case_sensitive_path(tmp_path, path, existing_paths, expected):
 
     result = find_case_sensitive_path(tmp_path / path, sys.platform)
     assert result == tmp_path / expected
+
+
+@pytest.fixture()
+def simple_module(tmp_path: Path) -> Path:
+    fn = tmp_path / "_src/tests/mymod.py"
```
Comment (NickCrews) [inline tests/test_path.py:130]: tests -> tasks

File: tests/test_path.py [hunk 4]
```diff
@@ -117,3 +123,179 @@ def test_find_case_sensitive_path(tmp_path, path, existing_paths, expected):
...
+
+    monkeypatch.setattr(
+        importlib.util, "spec_from_file_location", lambda *args: None  # noqa: ARG005
+    )
+    with pytest.raises(ImportError):
+        import_path(simple_module, root=tmp_path)
+
+
+def test_importmode_importlib_with_dataclass(tmp_path: Path) -> None:
+    """Ensure that importlib mode works with a module containing dataclasses (#7856)."""
```
Comment (NickCrews) [inline tests/test_path.py:175]: swap issue to https://github.com/pytask-dev/pytask/pull/373

Comment (NickCrews) [reply tests/test_path.py:175]: Or point to the absolute pytest issue https://github.com/pytest-dev/pytest/issues/7856

File: tests/test_path.py [hunk 5]
```diff
@@ -117,3 +123,179 @@ def test_find_case_sensitive_path(tmp_path, path, existing_paths, expected):
...
+
+    module = import_path(fn, root=tmp_path)
+    Data: Any = module.Data  # noqa: N806
+    data = Data(value="foo")
+    assert data.value == "foo"
+    assert data.__module__ == "_src.tests.test_dataclass"
+
+
+def test_importmode_importlib_with_pickle(tmp_path: Path) -> None:
+    """Ensure that importlib mode works with pickle (#7859)."""
```
Comment (NickCrews) [inline tests/test_path.py:198]: same as above, point to specific issue

File: tests/test_path.py [hunk 6]
```diff
@@ -117,3 +123,179 @@ def test_find_case_sensitive_path(tmp_path, path, existing_paths, expected):
...
+    fn = tmp_path.joinpath("_src/tests/test_pickle.py")
```
Comment (NickCrews) [inline tests/test_path.py:199]: test -> task

File: tests/test_collect.py [hunk 7]
```diff
@@ -37,6 +37,28 @@ def task_write_text(depends_on, produces):
     assert tmp_path.joinpath("out.txt").read_text() == "Relative paths work."
 
 
+@pytest.mark.end_to_end()
+def test_collect_module_name_(tmp_path):
```
Comment (NickCrews) [inline tests/test_collect.py:41]: I think we need a test that task modules can have the same name, eg `one/task_a.py` and `two/task_a.py`. I looked through this file (test_collect.py`) and I didn't see a test that did this, but maybe I'm missing it somewhere.

**Comment by tobiasraabe** [comment]
Hi @NickCrews, thank you for your review 🙏. Unfortunately, I won't be able to answer you soon, but I will return to you on Monday.

**Comment by tobiasraabe** [comment]
Thanks for staying with me!

> You're saying this is the drawback of prepend and append, right? Because this is NOT a requirement for importlib, what you have implemented here. I think we should add a test for this.

You are right. I misremembered when I wrote it.

> I respectfully disagree. I would like to have my tasks mixed with my "main logic" code...

I should have been more precise since I do not disagree with your setup. Nor do I want to strictly separate app and task code. I was thinking more about situations like those described in this [article of the documentation](https://pytask-dev.readthedocs.io/en/stable/how_to_guides/bp_scalable_repetitions_of_tasks.html).

In this example, general functions that provide access to the inputs and outputs of a stage like `data_preparation` are stored in `config.py` files. Then, the following step, `analysis` can access the results of the previous stage by importing `path_to_processed_data` from `my_project.data_preparation.config` instead of `my_project.data_preparation.task_prepare_data`.

You don't need to do that, as you said. Though, in bigger projects, this recommendation has helped beginners to care more about code separation, avoiding long modules, and by that also avoiding circular imports.

It might be misguided to say it is best practice because it concerns this specific case and is only documented in one article.

I will integrate the proposed changes soon and, then, I will release a new version as well.

File: tests/test_collect.py (hunk 7 above)
Comment (tobiasraabe) [reply tests/test_collect.py]: Good idea! I guess this case is only indirectly covered by some other test and it is better to have it explicit. Added as `test_collect_tasks_from_modules_with_the_same_name`.

**Comment by NickCrews** [comment]
This looks great, thank you so much for your work here!

**Pull request merged**